### DISCUSSION TITLE
The official example lacks interaction

### DISCUSSION THREAD

**Discussion by junmaqiang** [fyne-io/fyne#3572] [Q&A] [answered]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

![WX20230114-205731](https://user-images.githubusercontent.com/1819457/212472747-ea9326a7-0098-48d4-9769-49ca9ecb9eb2.png)

Missing how to interact, how to CURD

### Describe the solution you'd like to see.

It should be added to the operation item, CURD and other interactions

**Comment by SKBotNL** [comment]
It's supposed to be like that, it's a demo. If you want to know more about tables go to the [documentation](https://developer.fyne.io/widget/table)

**Comment by SKBotNL** [comment]
What doesn't work/what error do you get?

**Comment by Jacalz** [comment]
Please @junmaqiang, use a discussion for questions. This is an issue tracker and intended for reporting bugs and feature requests. I will convert this for you but please use https://github.com/fyne-io/fyne/discussions for questions next time.

I suppose you mean CRUD? If you need to remove or delete items just make sure that the `Length` function (the first passed to `NewTable`) returns the length of a data source instead of hard-coding it to numbers like in the example above. Then you can just remove the item from the slice and call `.Refresh()` on the table.

**Reply by junmaqiang** [reply]
> Please @junmaqiang, use a discussion for questions. This is an issue tracker and intended for reporting bugs and feature requests. I will convert this for you but please use https://github.com/fyne-io/fyne/discussions for questions next time.
> 
> I suppose you mean CRUD? If you need to remove or delete items just make sure that the `Length` function (the first passed to `NewTable`) returns the length of a data source instead of hard-coding it to numbers like in the example above. Then you can just remove the item from the slice and call `.Refresh()` on the table.

### Ok, thanks for the reminder. 

### My question is to know how to pass parameters to the method of the component?

**Reply by junmaqiang** [reply]
f.OnActivated = getInfoa(id.Row)  //  My wording is incorrect

**Reply by Jacalz** [reply]
You cannot set the value of the `OnActivated` field to the return value of a function that doesn't return anything. If you want to give it a function then you need to pass the function itself and not call it (i.e. set it to the name of the function).

However,I assume you want to do this instead (the types of functions are different):
```go
f.OnActivated = func() {
    getInfoa(id.Row)
}
```

**Comment by bq-wrongway** [comment]
hey, i think you can take a look at this, and i will also provide you with the working code.
https://fynelabs.com/2022/12/30/building-complex-tables-with-fyne/

```
package main

import (
	"fmt"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Devices")
	w.Resize(fyne.NewSize(320, 240))

	t := widget.NewTable(func() (int, int) {
		return 100, 4
	}, func() fyne.CanvasObject {
		return container.NewMax(widget.NewLabel("template11"), widget.NewButtonWithIcon("", nil, nil))
	}, func(id widget.TableCellID, o fyne.CanvasObject) {
		l := o.(*fyne.Container).Objects[0].(*widget.Label)
		i := o.(*fyne.Container).Objects[1].(*widget.Button)

		i.OnTapped = func() {
			fmt.Println(id.Row)

		}
		l.Show()
		i.Hide()

		switch id.Col {
		case 2:
			l.Hide()
			i.Show()

			i.SetIcon(getIcon(id.Row))
		case 0:
			l.SetText("hostname")
		case 1:
			l.SetText("127.0.0.1")
		case 3:
			l.SetText("notes...")
		}

	})
	t.SetColumnWidth(2, 24)
	t.SetColumnWidth(3, 156)

	w.SetContent(t)

	w.ShowAndRun()
}

func getIcon(i int) fyne.Resource {
	switch i % 3 {
	case 1:
		return theme.HomeIcon()
	case 2:
		return theme.MailSendIcon()
	default:
		return theme.MediaVideoIcon()
	}
}
```

**Reply by junmaqiang** [reply]
this all codes

```golang
package main

import (
	"fmt"
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
	"os"
)

var MoneyData = [][]string{
	{"218991153800884", "2022-11-12 10:23:01", "job", "12", "20", "A"},
	{"218991153800884", "2022-01-12 10:23:11", "cron", "12", "20", "B"},
	{"218991153800884", "2022-12-12 10:23:12", "pay", "-12", "20", "C"},
	{"218991153800884", "2022-09-12 10:13:01", "move", "-2", "20", "D"},
}

func main() {
	a := app.NewWithID("demo.io.fyne")
	a.SetIcon(theme.HomeIcon())
	w := a.NewWindow("Demo for tables")
	w.SetPadded(false)
	w.Resize(fyne.NewSize(1078, 598))
	t := widget.NewTable(
		func() (int, int) { return 4, 8 },
		func() fyne.CanvasObject {
			return container.NewMax(
				widget.NewLabel("name"),
				widget.NewToolbar(
					widget.NewToolbarSpacer(),
					widget.NewToolbarAction(
						theme.MailComposeIcon(),
                                                 
						func() { fmt.Println("Click") }, // I want this can use id.Row
						// func(id int){ fmt.Println("Row num:%d",id)} // That's what I wanted
					)))
		},
		func(id widget.TableCellID, cell fyne.CanvasObject) {
			label := cell.(*fyne.Container).Objects[0].(*widget.Label)
			icon := cell.(*fyne.Container).Objects[1].(*widget.Toolbar)
			//f := cell.(*fyne.Container).Objects[1].(*widget.Toolbar).Items[1].(*widget.ToolbarAction)
			label.Show()
			icon.Hide()
			switch id.Col {
			case 0:
				label.SetText(fmt.Sprintf("%d", id.Row+1))
			case 1:
				label.SetText(fmt.Sprintf(MoneyData[id.Row][id.Col-1]))
			case 2:
				label.SetText(fmt.Sprintf(MoneyData[id.Row][id.Col-1]))
			case 3:
				label.SetText(fmt.Sprintf(MoneyData[id.Row][id.Col-1]))
				label.Alignment = fyne.TextAlignCenter
			case 4:
				label.SetText(fmt.Sprintf("¥%s", MoneyData[id.Row][id.Col-1]))
				label.Alignment = fyne.TextAlignTrailing
			case 5:
				label.SetText(fmt.Sprintf("¥%s", MoneyData[id.Row][id.Col-1]))
				label.Alignment = fyne.TextAlignTrailing
			case 6:
				label.SetText(fmt.Sprintf(MoneyData[id.Row][id.Col-1]))
				label.Alignment = fyne.TextAlignCenter
			case 7:
				label.Hide()
				icon.Show()
				//f()
				//f = getInfo(id.Row)
				//f.OnActivated = getInfo()
				//f.OnActivated = getInfoa(id.Row)
			}
		})
	t.SetColumnWidth(0, 54)
	t.SetColumnWidth(1, 140)
	t.SetColumnWidth(2, 150)
	t.SetColumnWidth(3, 90)
	t.SetColumnWidth(4, 120)
	t.SetColumnWidth(5, 140)
	t.SetColumnWidth(6, 80)
	t.SetColumnWidth(7, 130)
	w.SetContent(t)
	w.ShowAndRun()
}
```

**Comment by andydotxyz** [comment, marked as answer]
Maybe you want the `OnSelected` field? If you set that to a func of your code it will pass the id that is being selected.

**Reply by junmaqiang** [reply]
yes, i want selected id.Row

**Reply by andydotxyz** [reply]
The demo you pointed at as lacking this behaviour absolutely does have that code: https://github.com/fyne-io/fyne/blob/8c2509518b2df442a6b748d9b07754739592e6d7/cmd/fyne_demo/main.go#L222.

**Reply by junmaqiang** [reply]
```golang
type ToolbarAction struct {
	Icon        fyne.Resource
	OnActivated func() `json:"-"`
}
```

how do I pass parameters to the method OnActivated

**Reply by Jacalz** [reply]
> how do I pass parameters to the method OnActivated

https://github.com/fyne-io/fyne/discussions/3572#discussioncomment-4686253

**Comment by junmaqiang** [comment]
This problem is solved
```golang
				icon.Items[1].(*widget.ToolbarAction).OnActivated = func() {
					fmt.Println("AA", id.Row)
				}

```

**Reply by andydotxyz** [reply]
I just wanted to mention that putting a Toolbar inside a table seems strange. You would normally have only 1 toolbar, and you can place it above table with `container.NewBorder(...)`.

**Reply by junmaqiang** [reply]
Used as a button, because there are icons and events

**Reply by Jacalz** [reply]
What is wrong with using a button? You can set `.Importance = widget.LowImportance` for it and also set an icon.

**Reply by junmaqiang** [reply]
I'm going to try it later

**Reply by junmaqiang** [reply]
fyne's table has no header which is very bad
